### PR TITLE
Implement Suggest Topics

### PR DESCRIPTION
**Pull request by tarassito**
## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/secure/RapidBoard.jspa?rapidView=777)

## Code reviewers

- [ ] @ashatna
- [ ] @ChristinaPastushchak
- [ ] @dotiurin
- [ ] @iryn34
- [ ] @tarassito
- [ ] @VasilHamulyak
- [x] @VitalyiKhlivnyi
- [ ] @YurijH

### Second Level Review

- [ ] @ushevtsiv
- [x] @lhalam
- [ ] @ssOleg

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: eventually/static/src/containers/suggestedtopics/InterestedUsersChips.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+const chipStyle = {
+    margin: 4,
+};
+
+class InterestedUsersChips extends React.Component {
+
+    constructor(props) {
+        super(props);
+    }
+    handleClick = () => {
```
Comment (VitalyiKhlivnyi) [inline eventually/static/src/containers/suggestedtopics/InterestedUsersChips.js]: add empty line between functions

File: eventually/static/src/containers/suggestedtopics/SuggestedTopicCreate.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                    onClick={this.handleOpen}
+                    style={FlatButtonStyle}>
+                    <ContentAdd />
+                </FloatingActionButton>
+                <Dialog
+                    actions={actions}
+                    modal={false}
+                    open={this.state.open}
+                    onRequestClose={this.handleClose}
+                    style={{zIndex: 800}}>
```
Comment (VitalyiKhlivnyi) [inline eventually/static/src/containers/suggestedtopics/SuggestedTopicCreate.js]: don`t use inline style

File: eventually/static/src/containers/suggestedtopics/SuggestedTopics.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                            this.setState(newUser);
+                        }
+                    });
+                }
+            }
+        });
+        this.setState({'suggestedTopics': suggestedTopics});
+    };
+
+
```
Comment (VitalyiKhlivnyi) [inline eventually/static/src/containers/suggestedtopics/SuggestedTopics.js]: delete empty line

File: eventually/static/src/containers/suggestedtopics/SuggestedTopicsItem.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        if (getUserId() == this.state.owner){
+            return true;
+        } else {
+            return false;
+        }
+    };
+
+    /*switch button from 'Join' to 'Leave' and backwards */
+    switchJoinButton = () => {
+        if (this.state.removeInterest === '0'){
```
Comment (VitalyiKhlivnyi) [inline eventually/static/src/containers/suggestedtopics/SuggestedTopicsItem.js]: change '0/1' in removeInterest variable to 'false/true'

File: eventually/suggestedtopics/models.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        :type name: string
+
+        :param description: Describing suggested topic
+        :type description: string
+
+        :param interested_user: List of users id interested in topic
+        :type interested_user: list of int
+
+        :param remove_interest: User id uninterested in topic
+        :type remove_interest: int
```
Comment (VitalyiKhlivnyi) [inline eventually/suggestedtopics/models.py]: change int to bool if you change remove_interest type

File: eventually/static/src/containers/suggestedtopics/InterestedUsersChips.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+                    onClick={this.handleClick}
+                    style={chipStyle}>
+                    {this.props.text}
+                </Chip>
+            </div>
+        );
+    }
+}
+
+export default withRouter(InterestedUsersChips);
```
Comment (ashatna) [inline eventually/static/src/containers/suggestedtopics/InterestedUsersChips.js:33]: Is there a new line in the end of document?

File: eventually/static/src/containers/suggestedtopics/SuggestedTopicCreate.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    handleSubmit = event => {
+        const data = {
+            'name': this.state.name,
+            'description': this.state.description
+        };
+        postSuggestedTopicsItem(data).then(response => {
+            this.props.getSuggestedTopicsItem();
+            this.handleClose();
+        });
+    };
```
Comment (ashatna) [inline eventually/static/src/containers/suggestedtopics/SuggestedTopicCreate.js]: You give possibility to save topic even it has no name or description

File: eventually/static/src/containers/suggestedtopics/SuggestedTopicCreate.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
+import React from 'react';
+import Dialog from 'material-ui/Dialog';
+import FlatButton from 'material-ui/FlatButton';
+import RaisedButton from 'material-ui/RaisedButton';
```
Comment (ashatna) [inline eventually/static/src/containers/suggestedtopics/SuggestedTopicCreate.js]: do you need RaisedButton import here?

File: eventually/static/src/containers/suggestedtopics/SuggestedTopicsEdit.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    /*save info from dialog*/
+    handleSave = event => {
+        const name = this.state.name;
+        const description = this.state.description;
+        const id = this.props.id;
+        putSuggestedTopicsItem(id, name, description).then(response => {
+            this.props.updateTopic(id, name, description);
+            this.props.handleClose();
+        });
+    };
```
Comment (ashatna) [inline eventually/static/src/containers/suggestedtopics/SuggestedTopicsEdit.js]: You give possibility to save topic even it has no name or description. And there is no check on the correctness of entering a name and description

File: eventually/suggestedtopics/models.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        | }
+        """
+
+        interested_users = [
+            user.id for user in self.interested_users.all()
+        ] if self.interested_users else []
+        interested_users_name = [
+            user.first_name + ' ' + user.last_name for user in self.interested_users.all()
+        ] if self.interested_users else []
+
```
Comment (ashatna) [inline eventually/suggestedtopics/models.py]: square brackets зщішешщт are not in accordance with the convention

File: eventually/suggestedtopics/models.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+        return {'id': self.id,
+                'name': self.name,
+                'description': self.description,
+                'owner': self.owner.id,
+                'interested_users': interested_users,
+                'interested_users_name': interested_users_name,
+                'created_at': int(self.created_at.timestamp()),
+                'updated_at': int(self.updated_at.timestamp())
+               }
```
Comment (ashatna) [inline eventually/suggestedtopics/models.py]: brackets are not in accordance with the convention

File: eventually/suggestedtopics/views.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        data = request.body
+
+        if not data:
+            return RESPONSE_400_EMPTY_JSON
+        keys_required = ['name', 'description']
+        if not required_keys_validator(data, keys_required, False):
+            return RESPONSE_400_INVALID_DATA
+        suggested_topic = SuggestedTopics.create(author, **data)
+        if suggested_topic and required_keys_validator(data, keys_required, False):
+            return JsonResponse(suggested_topic.to_dict(), status=201)
```
Comment (ashatna) [inline eventually/suggestedtopics/views.py]: you don't need second 'if required_keys_validator' on line 68

File: eventually/suggestedtopics/views.py [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+                 400, 403 or 404 failed status code.
+        :rtype: `HttpResponse object."""
+
+        suggested_topic = SuggestedTopics.get_by_id(suggested_topic_id)
+
+        if not suggested_topic:
+            return RESPONSE_404_OBJECT_NOT_FOUND
+        data = request.body
+        suggested_topic.update(**data)
+        return RESPONSE_200_UPDATED
```
Comment (ashatna) [inline eventually/suggestedtopics/views.py]: data validation is absent

File: eventually/static/src/containers/suggestedtopics/InterestedUsersChips.js (hunk 6 above)
Comment (dotiurin) [reply eventually/static/src/containers/suggestedtopics/InterestedUsersChips.js:33]: yes

**Comment by lhalam** [comment]
rebase branch and fix

27.33s$ pylint --rcfile=.pylintrc *
************* Module suggestedtopics.models
C: 75, 0: Trailing whitespace (trailing-whitespace)
W: 79, 0: Bad indentation. Found 19 spaces, expected 16 (bad-indentation)
C: 79, 0: No space allowed before bracket
                   interested_users_name.append(user.first_name + ' ' + user.last_name )
                                                                                       ^ (bad-whitespace)
C:142,11: Comparison to False should be 'not expr' or 'expr is False' (singleton-comparison)
C:144,13: Comparison to True should be just 'expr' or 'expr is True' (singleton-comparison)

**Review by lhalam — DISMISSED** [review]

**Comment by lhalam** [comment]
This branch is out-of-date with the base branch
Merge the latest changes from dev into this branch.

**Review by VitalyiKhlivnyi — APPROVED** [review]

**Review by lhalam — APPROVED** [review]

**Review by lhalam — APPROVED** [review]

**Pull request merged**